src/k2/module/k23si/TxnManager.cpp

```diff
@@ -192,247 +199,479 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
+    auto action = request.action == dto::EndAction::Commit ? TxnRecord::Action::onCommit : TxnRecord::Action::onAbort;
+    if (request.action == dto::EndAction::None) {
+        K2LOG_D(log::skvsvr, "cannot end transaction with None action in request {}", request);
+        return RPCResponse(dto::K23SIStatus::BadParameter("cannot end transaction with `None` end action"), dto::K23SITxnEndResponse{});
+    }
+    // store the write keys into the txnrecord
+    TxnRecord& rec = getTxnRecord(dto::TxnId{.trh = std::move(request.key), .mtr = std::move(request.mtr)});
+    if (rec.finalizeAction != dto::EndAction::None) {
+        // the record indicates we've received an end request already (there is a finalize action)
+        // this is only possible if there is a retry or some client bug
+        K2LOG_D(log::skvsvr, "TxnEnd retry - transaction already has a finalize action {}", rec)
+        return _endTxnRetry(rec, std::move(request));
+    }
+    rec.writeKeys = std::move(request.writeKeys);
+    rec.syncFinalize = request.syncFinalize;
+    rec.timeToFinalize = request.timeToFinalize;
+    rec.finalizeAction = request.action;
+    rec.hasAttemptedCommit = request.action == dto::EndAction::Commit;
+
+    // and just execute the transition
+    return _onAction(action, rec)
+        .then([this] (auto&& status) {
+            if (!status.is2xxOK()) {
+                return RPCResponse(std::move(status), dto::K23SITxnEndResponse{});
+            }
+
+            return RPCResponse(dto::K23SIStatus::OK("transaction ended"), dto::K23SITxnEndResponse{});
+        });
+}
+
+seastar::future<std::tuple<Status, dto::K23SITxnEndResponse>>
+TxnManager::_endTxnRetry(TxnRecord& rec, dto::K23SITxnEndRequest&& request) {
+    K2LOG_D(log::skvsvr, "duplicate end request {}, have= {}", request, rec);
+    if (rec.finalizeAction != request.action ||
+        rec.writeKeys != request.writeKeys) { // TODO add keyHash computed by client and compare here to verify contents
+        K2LOG_D(log::skvsvr, "invalid txn end retry request: rec={}, request={}", rec, request);
+        return RPCResponse(dto::K23SIStatus::BadParameter("end request retry does not match previous end request"),dto::K23SITxnEndResponse{});
+    }
+
+    // Upon a retry, respond after the current bgtask queue has drained (potential in-flight retries)
+    seastar::promise<std::tuple<Status, dto::K23SITxnEndResponse>> prom;
+    auto fut = prom.get_future();
+    _addBgTask(rec,
+        [prom=std::move(prom), this, txnId=rec.txnId, request=std::move(request)] () mutable {
+            // after previous bg tasks completed, see if we've finalized
+            auto it = _transactions.find(txnId);
+            if (it == _transactions.end()) {
+                // normally, we won't find a record in memory after the bg tasks have been done
+                prom.set_value(std::tuple(dto::K23SIStatus::OK("transaction retry ended"), dto::K23SITxnEndResponse{}));
```

Comment:
I think we talked about this before, but is this correct? We don't actually know here if the transaction was successfully committed right?

Reply:
you're right. Fixed to just return a KEY_NOT_FOUND for now with a comment for future optimization